components/core/src/clp/hash_utils.cpp

```diff
@@ -0,0 +1,107 @@
+#include "hash_utils.hpp"
+
+#include <span>
+#include <string>
+#include <vector>
+
+#include <fmt/format.h>
+#include <openssl/hmac.h>
+#include <openssl/sha.h>
+
+#include "ErrorCode.hpp"
+#include "spdlog_with_specializations.hpp"
+
+using std::span;
+using std::string;
+using std::vector;
+
+namespace clp {
+auto EvpDigestContext::digest_update(std::span<unsigned char const> input) -> ErrorCode {
+    if (m_is_digest_finalized) {
+        return ErrorCode_Unsupported;
+    }
+    if (1 != EVP_DigestUpdate(m_md_ctx, input.data(), input.size())) {
+        return ErrorCode_Failure;
+    }
+    return ErrorCode_Success;
+}
+
+auto EvpDigestContext::digest_final_ex(std::vector<unsigned char>& hash) -> ErrorCode {
+    if (m_is_digest_finalized) {
+        return ErrorCode_Unsupported;
+    }
+
+    hash.resize(SHA256_DIGEST_LENGTH);
+    unsigned int length{};
+    if (1 != EVP_DigestFinal_ex(m_md_ctx, hash.data(), &length)) {
+        return ErrorCode_Failure;
+    }
+    if (SHA256_DIGEST_LENGTH != length) {
+        return ErrorCode_Corrupt;
+    }
+    m_is_digest_finalized = true;
```

Comment:
This is antithetical to RAII since after this method is called, this class is not in a valid/usable state. I would suggest that we call `EVP_DigestInit_Ex` to reinitialize the context rather than maintaining this bool.